src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -418,42 +418,43 @@ static inline void log_circularity_error(Symbol* name, PlaceholderEntry* probe)
                                                                     Symbol* next_name,
                                                                     Handle class_loader,
-                                                                    bool is_superclass,
+                                                                    bool check_is_superclass,
                                                                     TRAPS) {
 
   assert(next_name != nullptr, "null superclass for resolving");
   assert(!Signature::is_array(next_name), "invalid superclass name");
 
-  // If class_name is already loaded, just return the superclass or superinterface.
-  // Make sure there's a placeholder for the class_name before resolving.
-  // This is used as a claim that this thread is currently loading superclass/classloader
-  // and for ClassCircularity checks.
-
   ClassLoaderData* loader_data = class_loader_data(class_loader);
-  Dictionary* dictionary = loader_data->dictionary();
 
-  // can't throw error holding a lock
-  bool throw_circularity_error = false;
-  {
-    MutexLocker mu(THREAD, SystemDictionary_lock);
-    InstanceKlass* klassk = dictionary->find_class(THREAD, class_name);
+  // For RedefineClasses, the class is already loaded so don't create a placeholder for checking
+  // circularity error if the next_name matches class->super()->name() and if the class loaders match.
+  // Also rarely, if parallel loading completes loading at this point, we can also check the super class name.
+  if (check_is_superclass) {
+    InstanceKlass* klassk = loader_data->dictionary()->find_class(THREAD, class_name);
     InstanceKlass* quicksuperk;
-    // To support parallel loading: if class is done loading, just return the superclass
-    // if the next_name matches class->super()->name() and if the class loaders match.
-    if (klassk != nullptr && is_superclass &&
+    if (klassk != nullptr &&
        ((quicksuperk = klassk->java_super()) != nullptr) &&
        ((quicksuperk->name() == next_name) &&
          (quicksuperk->class_loader() == class_loader()))) {
       return quicksuperk;
-    } else {
-      // Must check ClassCircularity before checking if superclass is already loaded.
-      PlaceholderEntry* probe = PlaceholderTable::get_entry(class_name, loader_data);
-      if (probe && probe->check_seen_thread(THREAD, PlaceholderTable::DETECT_CIRCULARITY)) {
-          log_circularity_error(class_name, probe);
-          throw_circularity_error = true;
-      }
+    }
+  }
+
+  // can't throw error holding a lock
+  bool throw_circularity_error = false;
+  {
+    MutexLocker mu(THREAD, SystemDictionary_lock);
+
+    // Must check ClassCircularity before resolving next_name (superclass or interface).
+    PlaceholderEntry* probe = PlaceholderTable::get_entry(class_name, loader_data);
+    if (probe && probe->check_seen_thread(THREAD, PlaceholderTable::DETECT_CIRCULARITY)) {
```

Comment:
Nit pre-existing use of implicit boolean - suggestion `probe != nullptr && ...`

Reply:
fixed.